src/java.base/share/classes/jdk/internal/foreign/abi/s390/linux/LinuxS390CallArranger.java

```diff
@@ -0,0 +1,312 @@
+import static jdk.internal.foreign.abi.s390.S390Architecture.Regs.*;
+
+/**
+ * For the S390 C ABI specifically, this class uses CallingSequenceBuilder
+ * to translate a C FunctionDescriptor into a CallingSequence, which can then be turned into a MethodHandle.
+ *
+ * This includes taking care of synthetic arguments like pointers to return buffers for 'in-memory' returns.
+ */
+public class LinuxS390CallArranger {
+
+    private static final int STACK_SLOT_SIZE = 8;
+    public static final int MAX_REGISTER_ARGUMENTS = 5;
+    public static final int MAX_FLOAT_REGISTER_ARGUMENTS = 4;
+
+    private static final ABIDescriptor CLinux = abiFor(
+            new VMStorage[] { r2, r3, r4, r5, r6, }, // GP input
+            new VMStorage[] { f0, f2, f4, f6 }, // FP input
+            new VMStorage[] { r2, }, // GP output
+            new VMStorage[] { f0, }, // FP output
+            new VMStorage[] { r0, r1, r2, r3, r4, r5, r14 }, // volatile GP
+            new VMStorage[] { f1, f3, f5, f7 }, // volatile FP (excluding argument registers)
+            8, // Stack is always 8 byte aligned on S390
+            160, // ABI header
+            r0, r1 // scratch reg r0 & r1
+            );
+
+    public record Bindings(CallingSequence callingSequence, boolean isInMemoryReturn) {}
+
+    public static Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean forUpcall) {
+        return getBindings(mt, cDesc, forUpcall, LinkerOptions.empty());
+    }
+
+    public static Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean forUpcall, LinkerOptions options) {
+        CallingSequenceBuilder csb = new CallingSequenceBuilder(CLinux, forUpcall, options);
+
+        BindingCalculator argCalc = forUpcall ? new BoxBindingCalculator(true) : new UnboxBindingCalculator(true);
+        BindingCalculator retCalc = forUpcall ? new UnboxBindingCalculator(false) : new BoxBindingCalculator(false);
+
+        boolean returnInMemory = isInMemoryReturn(cDesc.returnLayout());
+        if (returnInMemory) {
+            Class<?> carrier = MemorySegment.class;
+            MemoryLayout layout =SharedUtils.C_POINTER;
+            csb.addArgumentBindings(carrier, layout, argCalc.getBindings(carrier, layout));
+        } else if (cDesc.returnLayout().isPresent()) {
+            Class<?> carrier = mt.returnType();
+            MemoryLayout layout = cDesc.returnLayout().get();
+            csb.setReturnBindings(carrier, layout, retCalc.getBindings(carrier, layout));
+        }
+
+        for (int i = 0; i < mt.parameterCount(); i++) {
+            Class<?> carrier = mt.parameterType(i);
+            MemoryLayout layout = cDesc.argumentLayouts().get(i);
+            csb.addArgumentBindings(carrier, layout, argCalc.getBindings(carrier, layout));
+        }
+
+        return new Bindings(csb.build(), returnInMemory);
+    }
+
+    public static  MethodHandle arrangeDowncall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
```

Comment:
```suggestion
    public static MethodHandle arrangeDowncall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
```

Reply:
Thank you @turbanoff for the review comment. I have Fixed in new commit.